File: chapters/tree_traversal/code/cs/Tree/Tree.cs
// submitted by Julian Schacher (jspp)
using System;
using System.Collections.Generic;

namespace TreeTraversal
{
    public class Tree
    {
        private class Node
        {
            public List<Node> Children { get; set; } = new List<Node>();
            public int Id { get; set; }
        }

        private Node root;

        public Tree(int depthCount, int childrenCount)
        {
            CreateTree(depthCount, childrenCount);
        }

        public void CreateTree(int depthCount, int childrenCount)
        {
            root = new Node
            {
                Id = 1
            };
            CreateAllChildren(root, depthCount, childrenCount);
        }

        public void StartDFSRecursive()
        {
            DFSRecursive(root);
        }

        public void DFSStack()
        {
            var stack = new Stack<Node>();
            stack.Push(root);
            Node temp;

            while (stack.Count != 0)
            {
                Console.WriteLine(stack.Peek().Id);
                temp = stack.Pop();

                foreach (var c in temp.Children)
                {
                    stack.Push(c);
                }
            }
        }

        public void BFSQueue()
        {
            var queue = new Queue<Node>();
            queue.Enqueue(root);
            Node temp;

            while (queue.Count != 0)
            {
                Console.WriteLine(queue.Peek().Id);
                temp = queue.Dequeue();

                foreach (var c in temp.Children)
                {
                    queue.Enqueue(c);
                }
            }
        }

        private void CreateAllChildren(Node node, int rowCount, int childrenCount)
        {
            if (rowCount <= 1)
                return;

            for (int i = 0; i < childrenCount; i++)
            {
                node.Children.Add(new Node
                {
                    Id = node.Id * 10 + i + 1
                });
                CreateAllChildren(node.Children[i], rowCount - 1, childrenCount);
            }
        }

        private void DFSRecursive(Node node)
        {
            Console.WriteLine(node.Id);

            foreach (var c in node.Children)
            {
                DFSRecursive(c);
            }
        }
    }
}

File: chapters/tree_traversal/code/java/demo_code/Tree.java
// submitted by xam4lor
import java.util.ArrayList;
import java.util.PriorityQueue;
import java.util.Queue;
import java.util.Stack;

public class Tree {
    private class Node implements Comparable<Node> {
        public ArrayList<Node> children;
        public int id;

        public Node() {
            this.children = new ArrayList<Node>();
        }

        @Override
        public int compareTo(Node other) {
            // Need to implement Comparable<Node> and override this
            // method because of the method BFSQueue() which uses Queues
            // and must know how to check if two nodes are the same or not
            return Integer.compare(this.id, other.id);
        }
    }


    // Global node
    private Node root;

    public Tree(int depthCount, int childrenCount) {
        this.CreateTree(depthCount, childrenCount);
    }




    public void StartDFSRecursive() {
        this.DFSRecursive(this.root);
    }

    public void DFSStack() {
        Stack<Node> stack = new Stack<Node>();
        stack.push(this.root);

        Node tmp;

        while(stack.size() != 0) {
            System.out.println(stack.peek().id);
            tmp = stack.pop();

            for (Node c : tmp.children) {
                stack.push(c);
            }
        }
    }

    public void BFSQueue() {
        Queue<Node> queue = new PriorityQueue<Node>();
        queue.add(this.root);

        Node temp;

        while(queue.size() != 0) {
            System.out.println(queue.peek().id);
            temp = queue.poll(); // return null if the queue is empty

            if(temp != null) {
                for (Node c : temp.children) {
                    queue.add(c);
                }
            }
        }
    }



    private void CreateTree(int depthCount, int childrenCount) {
        this.root = new Node();
        this.root.id = 1;

        this.CreateAllChildren(this.root, depthCount, childrenCount);
    }

    private void CreateAllChildren(Node node, int rowCount, int childrenCount) {
        if(rowCount <= 1)
            return;

        for (int i = 0; i < childrenCount; i++) {
            Node tmp = new Node();
            tmp.id = node.id * 10 + i + 1;

            node.children.add(tmp);

            this.CreateAllChildren(node.children.get(i), rowCount - 1, childrenCount);
        }
    }

    private void DFSRecursive(Node node) {
        System.out.println(node.id);

        for (Node c : node.children) {
            DFSRecursive(c);
        }
    }
}
